video.cpp

```diff
@@ -0,0 +1,85 @@
+#include "recorder.hpp"
+#include "video.hpp"
+
+#ifdef USE_OPENCV_VIDEO_RECORDING
+#include <opencv2/core.hpp>
+#include <opencv2/imgproc.hpp>
+#include <opencv2/videoio.hpp>
+#include <sstream>
+
+namespace recorder {
+namespace {
+std::unique_ptr<cv::VideoWriter> buildOpenCVVideoWriter(const std::string &path, float fps, const cv::Mat &modelFrame) {
+    assert(fps > 0.0);
+    assert(!path.empty());
+    const auto codec = cv::VideoWriter::fourcc('M', 'J', 'P', 'G');
+    // This is the only thing we can write without FFMPEG on Android
+    // The path name should end with .avi
+    const auto backend = cv::CAP_OPENCV_MJPEG;
+    const bool isColor = modelFrame.channels() > 1;
+    {
+        // OpenCV writer gives no errors even if it is unable to open file and
+        // write frames to it, so test writing by ourselves.
+        std::ofstream video(path);
+        assert(video.is_open() && "unable to open video file for writing");
+    }
+    // log_info("recording %s video stream to %s", isColor ? "color" : "gray", path.c_str());
+    auto writer = std::make_unique<cv::VideoWriter>(
+            path, backend, codec, fps, modelFrame.size(), isColor);
+    assert(writer && "failed to create video writer");
+    // TODO: set video quality
+    return writer;
+}
+
+std::string videoOutputPath(const std::string &prefix, int cameraInd) {
+    std::ostringstream oss;
+    oss << prefix;
+    if (cameraInd != 0) {
+        oss << (cameraInd + 1);
+    }
+    oss << ".avi"; // must be .avi so OpenCV can record this without FFMPEG
+    return oss.str();
+}
+
+struct VideoWriterImplementation : public VideoWriter {
+    const std::unique_ptr<cv::VideoWriter> writer;
+    cv::Mat outputFrame;
+
+    VideoWriterImplementation(std::unique_ptr<cv::VideoWriter> writer) : writer(std::move(writer)) {}
+
+    void write(const cv::Mat &frame) final {
+        if (frame.channels() == 4) {
+            cv::cvtColor(frame, outputFrame, cv::COLOR_BGRA2BGR);
```

Comment:
nitpicking: would this conversion be better suited to be on Android side?

Reply:
I think this was tested on Android, but I'm not sure if it's actually an Android-specific issue